File: frontend/src/types/design.ts
export interface BoardSpec {
    width_mm: number;
    height_mm: number;
    layers: 2 | 4 | 6;
    outline: "rect" | "rounded_rect";
}

export interface ComponentConstraint {
    fixed_pos?: { x: number; y: number };
    layer?: "F.Cu" | "B.Cu";
    keep_away_mm?: number;
}

export interface Component {
    ref: string;
    part_id: string;
    footprint: string;
    value: string;
    constraints?: ComponentConstraint;
}

export interface NetPin {
    ref: string;
    pin: string;
}

export interface Net {
    name: string;
    net_class: "power" | "signal" | "differential";
    pins: NetPin[];
}

export interface DesignRules {
    min_clearance_mm: number;
    min_trace_width_mm: number;
    min_via_drill_mm: number;
    copper_weight_oz: 1 | 2;
    impedance_controlled: boolean;
}

export interface RoutingGoals {
    minimize: ("wire_length" | "vias" | "crosstalk")[];
    protect_nets: string[];
    fill_copper: boolean;
}

export interface PCBDesignRequest {
    prompt: string;
    project_name: string;
    board: BoardSpec;
    components?: Component[];
    nets?: Net[];
    rules?: DesignRules;
    routing_goals?: RoutingGoals;
}

export interface DRCViolation {
    severity: "error" | "warning";
    rule: string;
    description: string;
    location?: { x: number; y: number };
}

export interface DRCReport {
    violations: DRCViolation[];
    passed: boolean;
}

export interface PCBDesignResponse {
    job_id: string;
    status: "queued" | "running" | "complete" | "error";
    result?: {
        kicad_pcb_url: string;
        netlist_url: string;
        drc_report: DRCReport;
        placement_score?: number;
        routing_completion?: number;
        placements?: any[];
        traces?: any[];
    };
    errors?: string[];
}

File: backend/models/design.py
from pydantic import BaseModel, Field
from typing import Literal, Optional
from uuid import UUID

class BoardSpec(BaseModel):
    width_mm: float = Field(..., gt=0, le=500)
    height_mm: float = Field(..., gt=0, le=500)
    layers: Literal[2, 4, 6] = 2
    outline: Literal["rect", "rounded_rect"] = "rect"

class ComponentConstraint(BaseModel):
    fixed_pos: Optional[dict] = None   # {"x": float, "y": float}
    layer: Optional[Literal["F.Cu", "B.Cu"]] = None
    keep_away_mm: Optional[float] = None

class Component(BaseModel):
    ref: str                           # "U1", "C3", "R12"
    part_id: str                       # MPN or internal alias
    footprint: str                     # "Package_SO:SOIC-8_3.9x4.9mm"
    value: str                         # "10uF", "ESP32-WROOM"
    constraints: Optional[ComponentConstraint] = None

class NetPin(BaseModel):
    ref: str
    pin: str

class Net(BaseModel):
    name: str
    net_class: Literal["power", "signal", "differential"] = Field("signal", alias="class")
    pins: list[NetPin]

    class Config:
        populate_by_name = True

class DesignRules(BaseModel):
    min_clearance_mm: float = 0.15
    min_trace_width_mm: float = 0.2
    min_via_drill_mm: float = 0.3
    copper_weight_oz: Literal[1, 2] = 1
    impedance_controlled: bool = False

class RoutingGoals(BaseModel):
    minimize: list[Literal["wire_length", "vias", "crosstalk"]] = ["wire_length"]
    protect_nets: list[str] = []
    fill_copper: bool = True

class PCBDesignRequest(BaseModel):
    prompt: str
    project_name: str
    board: BoardSpec
    components: list[Component] = []
    nets: list[Net] = []
    rules: DesignRules = DesignRules()
    routing_goals: RoutingGoals = RoutingGoals()

class DRCViolation(BaseModel):
    severity: Literal["error", "warning"]
    rule: str
    description: str
    location: Optional[dict] = None

class DRCReport(BaseModel):
    violations: list[DRCViolation] = []
    warnings: list[DRCViolation] = []
    passed: bool

class PCBDesignResponse(BaseModel):
    job_id: str
    status: Literal["queued", "running", "complete", "error", "partial"]
    result: Optional[dict] = None      # kicad_pcb_url, netlist_url, drc_report, placement_score, routing_completion
    errors: list[str] = []
